tmate-ssh-server.c

```diff
@@ -506,7 +518,50 @@ void tmate_ssh_server_main(struct tmate_session *session, const char *keys_dir,
 
 		ssh_bind_free(bind);
 
-		client_bootstrap(session);
+		client_bootstrap_cb(session,
+                             handle_role_daemon,
+                             handle_role_pty_client,
+                             handle_role_exec);
 		/* never reached */
 	}
 }
+
+void tmate_ssh_server_main(struct tmate_session *session, const char *keys_dir,
+                           const char *bind_addr, int daemon_port, int client_port)
+{
+	struct tmate_ssh_client *client = &session->ssh_client;
+	ssh_bind daemon_bind, client_bind;
+	pid_t pid;
+
+	tmate_catch_sigsegv();
+	signal(SIGCHLD, handle_sigchld);
+
+	daemon_bind = prepare_ssh(keys_dir, bind_addr, daemon_port);
+	client_bind = prepare_ssh(keys_dir, bind_addr, client_port);
+
+	client->session = ssh_new();
+	client->channel = NULL;
+	client->winsize_pty.ws_col = 80;
+	client->winsize_pty.ws_row = 24;
+	session->session_token = "init";
+
+	if (!client->session)
+		tmate_fatal("Cannot initialize session");
+
+    if ((pid = fork()) < 0)
+      tmate_fatal("Can't fork");
+
+    if (pid != 0) {
+      /* Parent process */
+      handle_ssh_connections(session, client, daemon_bind, client_bootstrap,
+                             true /* handle_role_daemon */,
+                             false /* handle_role_pty_client */,
+                             false /* handle_role_exec */);
+    }
+
+    /* Child process */
+    handle_ssh_connections(session, client, client_bind, client_bootstrap,
```

Comment:
minor: for clarity, please add an `else` (or `return handle_ssh_connection(...)` in the previous `if`). From the point of view of this function, we cannot tell if `handle_ssh_connection` will return one day or not.

**Edit:** but this code may disappear.